### DISCUSSION TITLE
Entity movements with WASD?

### DISCUSSION THREAD

**Discussion by Koboo** [Minestom/Minestom#829] [Help]
Hey everyone, 

Relatively simple question, but has anyone ever tried to control entities with WASD? I haven't found a way to listen for the SteerVehiclePacket in JavaDocs or on the wiki, unless you use Packeteer and listen for the packets yourself. 

Did I miss something or just not find the solution?

**Comment by MrlnHi** [comment]
You can use
```java
MinecraftServer.getPacketListenerManager().setListener()
```
to listen for the packet

**Comment by TogAr2** [comment]
You can use `Player#getVehicleInformation()` to get vehicle controls

**Comment by RinesThaix** [comment]
I have it implemented, but code is a bit messy (and is on kotlin).

1. Firstly disable client vehicle move packet, because we don't want it to mess with our own movement logic: `MinecraftServer.getPacketListenerManager().setListener(ClientVehicleMovePacket::class.java) { _, _ -> }`.
2. Then create new class and extend Entity, providing it with the entity type of your choice. Override it's `update(Long)` method and make it look like this:
```kotlin
        super.update(time)
        val info = owner.vehicleInformation
        if (info.shouldUnmount() || shouldUnmount(time)) {
            remove()
            return
        }
        if (!info.shouldJump() && info.forward == 0F && info.sideways == 0F) {
            if (hasNoGravity()) {
                move(time, Vec(0.0, -gravityAcceleration, 0.0))
            }
            return
        }
        val delta = Vec(
            info.forward.coerceIn(-1F, 1F).toDouble(),
            0.0,
            -info.sideways.coerceIn(-1F, 1F).toDouble()
        ).rotateFromView(owner.position)
            .withY(0.0)
            .normalize()
            .mul(entitySpeed.toDouble())
            .withY(if (info.shouldJump() && canJump(time)) gravityAcceleration * 2 else 0.0)
        if (delta.x().isNaN()) {
            return
        }
        move(time, delta)
```

`owner` is a Player and `move` in my code is kinda abstract, but for ground-only mounts it looks like this:

EntityAbstractMount
```kotlin
protected open fun move(time: Long, delta: Vec, physicsResult: PhysicsResult? = null) {
        if (delta.x().isNaN()) {
            return
        }
        val physicsResult = physicsResult ?: CollisionUtils.handlePhysics(this, delta)
        onGround = physicsResult.isOnGround
        val position = physicsResult.newPosition().withView(
            PositionUtils.getLookYaw(delta.x(), delta.z()),
            PositionUtils.getLookPitch(delta.x(), delta.y(), delta.z())
        )
        instance!!.loadChunk(position).thenRun { refreshPosition(position) }
    }
```

EntityGroundMount (extends previous one)
```kotlin
override fun move(time: Long, delta: Vec, physicsResult: PhysicsResult?) {
        if (physicsResult != null) {
            super.move(time, delta, physicsResult)
            return
        }
        var physicsResult = CollisionUtils.handlePhysics(this, delta)
        val d = if (
            !physicsResult.newVelocity().samePoint(delta) &&
            (onGround || instance!!.getBlock(position.withY { it - 1 }).isSolid)
        ) {
            val d = delta.withY { it + 1 }
            physicsResult = CollisionUtils.handlePhysics(this, d)
            d
        } else {
            delta
        }
        super.move(time, d, physicsResult)
    }
```

After that, all you need to do is to spawn your mount, setup it's instance and add a passenger. My code for this looks as follows):
```kotlin
val mount = EntityGroundMount(...)
mount.setInstance(player.instance!!, player.position)
mount.addPassenger(player)
```

**Comment by Koboo** [comment]
You really gave a perfect answer! Thank you for the detailed information and explanation. Whether Kotlin or Java should not be a problem :D

If I understand correctly, I can then with `extends EntityAbstractMount` theoretically also create further movement logic for e.g. flying entities. I will try to test this out today and create a PoC for myself.

Thank you very much again!
